### PR TITLE
add a meta registration for split_with_sizes

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135699

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -6510,6 +6510,30 @@ def meta__padded_dense_to_jagged_forward(
...
+    torch._check_with(
+        ValueError,
+        sum(split_sizes) == self.shape[dim],
+        lambda: f"Split sizes add up to {sum(split_sizes)} but got the tensor's size of {self.shape[dim]}",
+    )
+    result = []
+    for split_size in split_sizes:
+        shape = list(self.shape)
+        shape[dim] = split_size
+        result.append(self.new_empty(shape))
```
Comment (ezyang) [inline torch/_meta_registrations.py:6533]: oh no, you can't do this lol, you have to actually return views 😭

Comment (isuruf) [reply torch/_meta_registrations.py:6533]: Ah, that makes the difference.

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -6510,6 +6510,33 @@ def meta__padded_dense_to_jagged_forward(
...
+        sum(split_sizes) == self.shape[dim],
+        lambda: f"Split sizes add up to {sum(split_sizes)} but got the tensor's size of {self.shape[dim]}",
+    )
+    result = []
+    for split_size in split_sizes:
+        new_shape = list(self.shape)
+        new_shape[dim] = split_size
+        # We mimic narrow which calls _slice_in_dim_meta which in turn calls _slice_meta
+        # By doing that we avoid a lot of checks in slice_in_dim and slice
+        result.append(self.as_strided(new_shape, self.stride(), self.storage_offset()))
```
Comment (isuruf) [inline torch/_meta_registrations.py:6536]: @ezyang, does this look correct?

Comment (isuruf) [reply torch/_meta_registrations.py:6536]: Apparently not

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -6510,6 +6510,38 @@ def meta__padded_dense_to_jagged_forward(
...
+        new_shape = list(self.shape)
+        new_shape[dim] = split_size
+        # We mimic narrow which calls _slice_in_dim_meta which in turn calls _slice_meta
+        # By doing that we avoid a lot of checks in slice_in_dim and slice
+        offset = self.stride()[dim] * start_idx
+        result.append(
+            self.as_strided(new_shape, self.stride(), self.storage_offset() + offset)
+        )
+        start_idx += split_size
+    return result
```
Comment (isuruf) [inline torch/_meta_registrations.py:6542]: This can be the decomposition as well.

**Comment by isuruf** [comment]
Closing this in favour of https://github.com/pytorch/pytorch/pull/135728

**Pull request closed without merging**